### PR TITLE
nextcloud.generate.sh: update script

### PR DESCRIPTION
**Pull request by teto**
so that it can be run from any folder.
it's also now more portable as it makes sure to run with nixpkgs' sed and jq (which were assumed to be installed before this PR).

cc @onny 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by teto** [comment]
actually this can't be run anywhere because nc4nix just dumps its file in the current directory :s but it's an improvement. I also dont like that we call `rm`: what's the point of logs if we remove them ? instead we should run the script into a TMPDIR or gitignore the logs.

**Comment by onny** [comment]
Thank you for this PR :) I'm not experienced enough to tell if this approach is the prefered way of doing this in nixpkgs. Some parts look like good improvements but personally I'm not sure if it's good to split up the logic into separate files. For me this increases the complexity. Would be good to ask other contributors for review

**Comment by teto** [comment]
I would say this is a more robust approach as it makes sure everyone uses the same tools installed from the same nixpgs. As it stands if you dont have jq in PATH, the script fails. 

What I would like to is to run nc4nix in parallels for the different nextcloud package sets. I suppose I could run a for loop and update NEXTCLOUD_VERSIONS to contain only one version each time but this feels hackish.

File: pkgs/servers/nextcloud/packages/generate.sh [hunk 1]
```diff
@@ -1,14 +1,15 @@
 #!/usr/bin/env nix-shell
-#! nix-shell -I nixpkgs=../../../.. -i bash -p nc4nix
+#!nix-shell nc4nix-shell.nix -i bash
 
 set -e
 set -u
 set -o pipefail
 set -x
```
Comment (SuperSandro2000) [inline pkgs/servers/nextcloud/packages/generate.sh:7]:
```suggestion
set -eou pipefail
```
running scripts always with -x is usually very noisy

File: pkgs/servers/nextcloud/packages/generate.sh [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
 
-export NEXTCLOUD_VERSIONS=$(nix-instantiate --eval -E 'import ./nc-versions.nix {}' -A e)
+echo "versions: $NEXTCLOUD_VERSIONS"
```
Comment (SuperSandro2000) [inline pkgs/servers/nextcloud/packages/generate.sh:9]: Do we now need to manually set this?

File: pkgs/servers/nextcloud/packages/generate.sh [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
 
-APPS=`cat nextcloud-apps.json | jq -r '.[]' | sed -z 's/\n/,/g;s/,$/\n/'`
+APPS=$(jq -r '.[]' "$SCRIPT_FOLDER"/nextcloud-apps.json | sed -z 's/\n/,/g;s/,$/\n/')
 
-nc4nix -apps $APPS
-rm *.log
+# nc4nix needs NEXTCLOUD_VERSIONS
+nc4nix -apps "$APPS"
+rm ./*.log
```
Comment (SuperSandro2000) [inline pkgs/servers/nextcloud/packages/generate.sh:15]:
```suggestion
rm -f *.log
```
don't fail if somehow we didn't generate a log

File: pkgs/servers/nextcloud/packages/nc4nix-shell.nix [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{ pkgs ? import ../../../.. { }
+}:
+let
+  lib = pkgs.lib;
+  n = lib.mapAttrsToList (_: v: v.version) (
+      lib.filterAttrs (k: v: builtins.match "nextcloud[0-9]+" k != null && (builtins.tryEval v.version).success)
+    pkgs);
```
Comment (SuperSandro2000) [inline pkgs/servers/nextcloud/packages/nc4nix-shell.nix:7]: Can't we regex on the current dir or so? That would probably be faster.

**Review by SuperSandro2000 — COMMENTED** [review]
The commit title should be more like ``nextcloud: allow running generate.sh from repo root`` or so

I would like to see that you can just start the script and not need to open some custom nix shell.

**Pull request closed without merging**